utils/ofdns/ofdns.1
.\"
.\" Copyright (c) 2008-2026 Jonathan Schleifer <js@nil.im>
.\"
.\" All rights reserved.
.\"
.\" This program is free software: you can redistribute it and/or modify it
.\" under the terms of the GNU Lesser General Public License version 3.0 only,
.\" as published by the Free Software Foundation.
.\"
.\" This program is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
.\" version 3.0 for more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public License
.\" version 3.0 along with this program. If not, see
.\" <https://www.gnu.org/licenses/>.
.\"
.TH ofdns 1
.SH NAME
ofdns \- perform DNS lookups
.SH SYNOPSIS
.B ofdns
[\fBoptions\fR] \fIdomain1\fR [\fIdomain2\fR ...]
.SH DESCRIPTION
.B ofdns
is a program to perform DNS lookups.
.SH OPTIONS
.TP
.BR \fB\-c\fR " " \fIclass\fR ", " \fB\-\-class=\fIclass\fR
The DNS class to query (defaults to IN).
.TP
.BR \fB\-h\fR ", " \fB\-\-help\fR
Show the help.
.TP
.BR \fB\-s\fR " " \fIserver\fR ", " \fB\-\-server=\fIserver\fR
The server to query.
.TP
.BR \fB\-t\fR " " \fItype\fR ", " \fB\-\-type=\fItype\fR
The record type to query (defaults to AAAA and A, can be repeated).
.TP
.BR \fB\-\-tcp\fR
Force using TCP for the query.
.SH EXAMPLES
Query the \fBA\fR and \fBAAAA\fR record for \fBexample.com\fR:
.PP
	ofdns example.com
.PP
Query the \fBCNAME\fR record for \fBwww.example.com\fR:
.PP
	ofdns \-t CNAME example.com
.PP
Query \fB2001:db8::1\fR for the \fBAAAA\fR record for \fBexample.com\fR:
.PP
	ofdns \-s 2001:db8::1 \-t AAAA example.com


ChangeLog
Legend:
 * Changes of existing features or bug fixes
 + New features

This file only contains the most significant changes.

ObjFW 1.5.4 -> ObjFW 1.5.5, 2026-05-31

 * Fixes replacing a string within a range that does not start at 0.
 * Fixes -[OFData rangeOfData:options:range:] if the specified range does not
   start at 0.
 * Fixes -[OFMutableData subdataWithRange:].
 * Fixes a missing retain/release in OFConcreteMutableArray.
 * Fixes -[initWithContentsOfIRI:] being extremely slow for both OFString and
   OFData on some operating systems such as Windows.
 * Fixes wrong count after -[OFList removeAllObjects].
 * Fixes objc_setAssociatedObject() / objc_removeAssociatedObjects() releasing
   the object within the spinlock, which could potentially deadlock.
 * Fixes OFBlock calling a dispose handler that doesn't exist.
 * Fixes -[OFConcreteMutableArray removeAllObjects] not increasing the
   mutations.
 * Fixes Deflate64 with sequence lengths > 65535.
 * Robustness improvements in Deflate, LHA, Tar, ZIP, BMP, QOI and MessagePack.
 * Fixes a missing rethrow in OFOpenSSLTLSStream that could result in hangs.
 * Correctly handles `*` in format strings now.
 * OFDate rejects invalid dates with tm_mon >= 12 now.
 * OFXMLParser rejects numeric entities == 0 or > U+10FFFF now, as well as
   surrogates.
 * OFString rejects overlong UTF-8 encodings now.
 * Fixes a retain cycle and a missing rethrow in OFCFRunLoopKernelEventObserver.
 * Fixes the number of redirects accidentally being reset in OFHTTPClient.
 * OFHTTPClient is now stricter about when to include the Authorization header
   during redirects.
 * Fixes -[objectByParsingJSON] reporting the wrong line on parsing errors.
 * Fixes -[objectByParsingJSON] overallocating.
 * Fixes -[objectByParsingJSON] not enforcing a low surrogate after a high
   surrogate.
 * Changes default depth limit to 128 for both JSON and MessagePack.
 * Base64 now rejects spurious `=` in the middle of the string.
 * OFHTTPClient and OFHTTPServer now have more robust handling of
   Transfer-Encoding.
 * OFDNSResolver is more compliant with the RFC now.
 * MessagePack now rejects nanoseconds >= 1000000000 as per specification.
 * Fixes OFConcreteArray accepting `nil` as the first object in some cases.
 * Fixes OFMatrix4x4 jumping to the wrong label in inline assembly, causing the
   loop to always run at least once.
 * Fixes wrong initialization in OFColorSpace.
 * Fixes -[OFColor colorUsingColorSpace:] always using sRGB.
 * Fixes some colors comparing equal despite having a different hash in OFColor.
 * OFAllocObject() and class_createInstance() now reject a too big extra size.
 * Fixes objc_setPropertyStruct() using the wrong spinlock.
 * Fixes MIPS32 lookup assembly using the wrong load size.
 * Fixes PowerPC 64 forwarding assembly comparing with the wrong register size.
 * Fixes a typo that made ARM32 forwarding assembly for Apple not compile.
 * Fixes a typo that made compiling for Windows without atomic operations fail.
 * OFApplication now checks the return value of unveil().

ObjFW 1.5.3 -> ObjFW 1.5.4, 2026-05-23
 * Fixes a buffer overflow caused by integer promotion rules in
   OFBMPImageFormatHandler and OFQOIImageFormatHandler.

ObjFW 1.5.2 -> ObjFW 1.5.3, 2026-05-20
 * Fixes OFForwardingTests on Windows with GCC.
 * Fixes OFHTTPCookieManager which was completely broken.
 * OFHTTPServer refuses to send responses that contain a newline in the
   headers or the server name now.
 * Fixes setting the Amiga Comment to nil in OFMutableTarArchiveEntry.
 * Improves handling of truncated archives in OFLHAArchiveEntry.
 * OFThread no longer inherits the priority on AmigaOS.
 * Uses better source for randomness now to find a free port on platforms that
   can't bind to port 0.
 * ofarc now strips drives from the paths it stores inside the archive.
 * Fixes ofarc failing to create ZIP archives with specific paths on AmigaOS.
 * Fixes ofarc incorrectly storing Amiga Protection in ZIP archives.

ObjFW 1.5.1 -> ObjFW 1.5.2, 2026-05-08
 * Makes ObjFWConfig.cmake work on MSYS2.
 * ObjFWConfig.cmake now also adds the flags for OBJCXX.
 * OFFile no longer uses O_EXLOCK for "x" modes.
 * Fixes the `OF_AMD64` define on Windows.
 * ofhttp now removes the terminal's progress indicator on abort.

ObjFW 1.5 -> ObjFW 1.5.1, 2026-02-24
 * Fixes a regression introduced in 1.5 that made weak references deadlock.
 * Improves the unit test for weak references.

ObjFW 1.4.4 -> ObjFW 1.5, 2026-02-22
 + Adds OFColorSpace and adds support for color spaces to OFColor. Supports
   sRGB, linear sRGB, BT.709, Display-P3, linear Display-P3, BT.2020, linear
   BT.2020, Adobe RGB and linear Adobe RGB.
 + Adds OFImage to store an image in various pixel formats and color spaces.
 + Adds OFCanvas to create (draw) images, including bilinear image scaling and
   alpha blending in linear color space and pixel format and color space
   transformations when source and destination use a different pixel format
   and/or color space.
 + Adds OFImageFormatHandler as a generic interface to handle image formats
   including a registry for image formats so that it can be extended by 3rd
   parties.
 + Adds an OFImageFormatHandler for BMP and QOI images.
 + Adds support for CMake projects using ObjFW.
 + Adds support for Unicode 17.
 + Adds methods to OFString and OFMutableString to remove control characters.
 + OFLocale now escapes all control characters in variables unless the
   variable name starts with an `@`.
 * OFStdOut and OFStdErr now allow writing with a lossy encoding if they have
   an underlying terminal.
 + OFStdOut and OFStdErr now support setting the terminal's progress
   indicator.
 * Revives support for AmigaOS 4 that had bitrotten over the years.
 + OFFileIRIHandler now supports nanosecond precision on operating systems
   that support it.
 + OFFileIRIHandler now supports setting file owner and group on UNIX systems.
 + OFFileIRIHandler now supports retrieving and setting of MS-DOS attributes,

utils/ofdns/OFDNS.m
/*
 * Copyright (c) 2008-2026 Jonathan Schleifer <js@nil.im>
 *
 * All rights reserved.
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License version 3.0 only,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
 * version 3.0 for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * version 3.0 along with this program. If not, see
 * <https://www.gnu.org/licenses/>.
 */

#include "config.h"

#import "OFApplication.h"
#import "OFArray.h"
#import "OFDNSResolver.h"
#import "OFData.h"
#import "OFIRI.h"
#import "OFLocale.h"
#import "OFOptionsParser.h"
#import "OFSandbox.h"
#import "OFStdIOStream.h"
#import "OFSystemInfo.h"
#import "OFThread.h"

#import "OFInvalidFormatException.h"

#ifdef OF_AMIGAOS
const char *VER = "$VER: ofdns " OF_PREPROCESSOR_STRINGIFY(OBJFW_VERSION_MAJOR)
    "." OF_PREPROCESSOR_STRINGIFY(OBJFW_VERSION_MINOR) " (" BUILD_DATE ") "
    "\xA9 2008-2026 Jonathan Schleifer";
#endif

@interface OFDNS: OFObject <OFApplicationDelegate, OFDNSResolverQueryDelegate>
{
	size_t _inFlight;
	int _errors;
}
@end

OF_APPLICATION_DELEGATE(OFDNS)

static void
help(OFStream *stream, bool full, int status)
{
	[OFStdErr writeLine: OF_LOCALIZED(@"usage",
	    @"Usage: %[prog] -[chst] domain1 [domain2 ...]",
	    @"prog", [OFApplication programName])];

	if (full) {
		[stream writeString: @"\n"];
		[stream writeLine: OF_LOCALIZED(@"full_usage",
		    @"Options:\n    "
		    @"-c  --class= "
		    @"  The DNS class to query (defaults to IN)\n    "
		    @"-h  --help   "
		    @"  Show this help\n    "
		    @"-s  --server="
		    @"  The server to query\n    "
		    @"-t  --type=  "
		    @"  The record type to query (defaults to AAAA and A,\n"
		    @"                   can be repeated)\n    "
		    @"    --tcp    "
		    @"  Force using TCP for the query\n    "
		    @"    --version"
		    @"  Print the version information")];
	}

	[OFApplication terminateWithStatus: status];
}

static void
version(void)
{
	[OFStdOut writeFormat:
	    @"ofdns %@ (ObjFW %@) "
	    @"<https://objfw.nil.im> <gemini://objfw.nil.im>\n"
	    @"Copyright (c) 2008-2026 Jonathan Schleifer <js@nil.im>\n"
	    @"Licensed under the LGPL 3.0 "
	    @"<https://www.gnu.org/licenses/lgpl-3.0.html>\n",
	    @PACKAGE_VERSION, [OFSystemInfo ObjFWVersion]];
	[OFApplication terminate];
}

@implementation OFDNS
-  (void)resolver: (OFDNSResolver *)resolver
  didPerformQuery: (OFDNSQuery *)query
	 response: (OFDNSResponse *)response
	exception: (id)exception
{
	_inFlight--;

	if (exception == nil)
		[OFStdOut writeFormat: @"%@\n", response];
	else {
		[OFStdErr writeLine: OF_LOCALIZED(@"failed_to_resolve",
		    @"Failed to resolve: %[exception]",
		    @"exception", exception)];
		_errors++;
	}

	if (_inFlight == 0)
		[OFApplication terminateWithStatus: _errors];
}

- (void)applicationDidFinishLaunching: (OFNotification *)notification
{
	OFString *DNSClassString, *server;
	bool forceTCP;
	const OFOptionsParserOption options[] = {
		{ 'c',  @"class", 1, NULL, &DNSClassString },
		{ 'h',  @"help", 0, NULL, NULL },
		{ 's',  @"server", 1, NULL, &server },
		{ 't',  @"type", 1, NULL, NULL },
		{ '\0', @"tcp", 0, &forceTCP, NULL },
		{ '\0', @"version", 0, NULL, NULL },
		{ '\0', nil, 0, NULL, NULL }
	};
	OFMutableArray OF_GENERIC(OFString *) *recordTypes;
	OFOptionsParser *optionsParser;
	OFUnichar option;
	OFArray OF_GENERIC(OFString *) *remainingArguments;
	OFDNSResolver *resolver;
	OFDNSClass DNSClass;

#ifdef OF_HAVE_FILES
# ifndef OF_AMIGAOS
	[OFLocale addLocalizationDirectoryIRI:
	    [OFIRI fileIRIWithPath: @LOCALIZATION_DIR]];
# else
	[OFLocale addLocalizationDirectoryIRI:
	    [OFIRI fileIRIWithPath: @"PROGDIR:/Data/OFDNS/localization"]];
# endif
#endif

#ifdef OF_HAVE_SANDBOX
	OFSandbox *sandbox = [[OFSandbox alloc] init];
	@try {
		sandbox.allowsStdIO = true;
		sandbox.allowsDNS = true;

		[OFApplication of_activateSandbox: sandbox];
	} @finally {
		objc_release(sandbox);
	}
#endif

	recordTypes = [OFMutableArray array];

	optionsParser = [OFOptionsParser parserWithOptions: options];
	while ((option = [optionsParser nextOption]) != '\0') {
		switch (option) {
		case 't':
			[recordTypes addObject: optionsParser.argument];
			break;
		case 'h':
			help(OFStdOut, true, 0);
			break;
		case '-':
			if ([optionsParser.lastLongOption isEqual: @"version"])
				version();
			break;
		case ':':
			if (optionsParser.lastLongOption != nil)
				[OFStdErr writeLine: OF_LOCALIZED(
				    @"long_option_required_argument",
				    @"%[prog]: Option --%[opt] requires an "
				    @"argument",
				    @"prog", [OFApplication programName],
				    @"opt", optionsParser.lastLongOption)];
			else {
				OFString *optStr = [OFString
				    stringWithFormat: @"%C",
				    optionsParser.lastOption];
				[OFStdErr writeLine: OF_LOCALIZED(
				    @"option_requires_argument",
				    @"%[prog]: Option -%[opt] requires an "
				    @"argument",
				    @"prog", [OFApplication programName],
				    @"opt", optStr)];
			}

			[OFApplication terminateWithStatus: 1];
			break;
		case '?':
			if (optionsParser.lastLongOption != nil)
				[OFStdErr writeLine: OF_LOCALIZED(
				    @"unknown_long_option",
				    @"%[prog]: Unknown option: --%[opt]",
				    @"prog", [OFApplication programName],
				    @"opt", optionsParser.lastLongOption)];
			else {
				OFString *optStr = [OFString
				    stringWithFormat: @"%C",
				    optionsParser.lastOption];
				[OFStdErr writeLine: OF_LOCALIZED(
				    @"Unknown_option",
				    @"%[prog]: Unknown option: -%[opt]",
				    @"prog", [OFApplication programName],
				    @"opt", optStr)];
			}

			[OFApplication terminateWithStatus: 1];
			break;
		}
	}

	remainingArguments = optionsParser.remainingArguments;

	if (remainingArguments.count < 1)
		help(OFStdErr, false, 1);

	resolver = [OFDNSResolver resolver];
	resolver.configReloadInterval = 0;
	resolver.forcesTCP = forceTCP;

	DNSClass = (DNSClassString != nil
	    ? OFDNSClassParseName(DNSClassString) : OFDNSClassIN);

	if (recordTypes.count == 0) {
		[recordTypes addObject: @"AAAA"];
		[recordTypes addObject: @"A"];
	}

	if (server != nil) {
		@try {
			OFSocketAddressParseIP(server, 0);

			resolver.nameServers =
			    [OFArray arrayWithObject: server];
		} @catch (OFInvalidFormatException *e) {
			/* Not an IP. Try resolving. */
			OFData *addressesData = [[OFThread DNSResolver]
			    resolveAddressesForHost: server
				      addressFamily: OFSocketAddressFamilyAny];
			const OFSocketAddress *addresses = addressesData.items;
			size_t count = addressesData.count;
			OFMutableArray *servers = [OFMutableArray
			    arrayWithCapacity: count];

			for (size_t i = 0; i < count; i++)
				[servers addObject:
				    OFSocketAddressString(&addresses[i])];

			[servers makeImmutable];
			resolver.nameServers = servers;
		}
	}

	for (OFString *domainName in remainingArguments) {
		for (OFString *recordTypeString in recordTypes) {
			OFDNSRecordType recordType =
			    OFDNSRecordTypeParseName(recordTypeString);
			OFDNSQuery *query =
			    [OFDNSQuery queryWithDomainName: domainName
						   DNSClass: DNSClass
						 recordType: recordType];

			_inFlight++;
			[resolver asyncPerformQuery: query delegate: self];
		}
	}
}
@end
